hummingbot/strategy/__utils__/trailing_indicators/trading_intensity.pyx

```diff
@@ -46,43 +47,81 @@ cdef class TradingIntensityIndicator:
 
     @property
     def is_sampling_buffer_full(self) -> bool:
-        return len(self._trade_samples) == self._sampling_length
+        return len(self._trade_samples.keys()) == self._sampling_length
 
     @property
     def is_sampling_buffer_changed(self) -> bool:
-        is_changed = self._samples_length != len(self._trade_samples)
-        self._samples_length = len(self._trade_samples)
+        is_changed = self._samples_length != len(self._trade_samples.keys())
+        self._samples_length = len(self._trade_samples.keys())
         return is_changed
 
     @property
     def sampling_length(self) -> int:
         return self._sampling_length
 
+    # A helper method to be used in unit tests
     @property
-    def last_price(self) -> float:
-        return self._last_price
+    def last_quotes(self) -> list:
+        return self._last_quotes
+
+    # A helper method to be used in unit tests
+    @last_quotes.setter
+    def last_quotes(self, value):
+        self._last_quotes = value
+
+    # A helper method to be used in unit tests
+    def calculate(self, timestamp):
+        self.c_calculate(timestamp)
+
+    cdef c_calculate(self, timestamp):
+        price = self._price_delegate.get_price_by_type(PriceType.MidPrice)
+        # Descending order of price-timestamp quotes
+        self._last_quotes = [{'timestamp': timestamp, 'price': price}] + self._last_quotes
+
+        trade_sample_rem = []
+        latest_processed_quote_idx = None
+        for trade in self._current_trade_sample:
+            is_processed = False
+            for i, quote in enumerate(self._last_quotes):
+                if quote["timestamp"] < trade.timestamp:
+                    is_processed = True
+                    if latest_processed_quote_idx is None or i < latest_processed_quote_idx:
+                        latest_processed_quote_idx = i
+                    trade = {"price_level": abs(trade.price - float(quote["price"])), "amount": trade.amount}
+
+                    if quote["timestamp"]+1 not in self._trade_samples.keys():
+                        self._trade_samples[quote["timestamp"]+1] = []
+
+                    self._trade_samples[quote["timestamp"]+1] += [trade]
+                    break
+            if not is_processed:
+                trade_sample_rem += [trade]
+
+        # Store unprocessed remainder for future processing
+        self._current_trade_sample = trade_sample_rem
```

Comment:
Would this ever be processed, since we're eliminating timestamps as we go along? I think the only scenario where this would happen is at the start of the run when we receive a trade that happened a second before we started the strategy. In that scenario, we don't have a ts to match it with, but we never will. If that's the case, then I think we can just discard those trades.

Reply:
Unlikely, but there's still a possibility that the snapshots will be delayed since there's no dependency between the snapshots and trade updates.

Reply:
The `timestamp` is taken as the current time stamp (i.e. the current seconds since epoch), which will guarantee that no cached trade will have a larger ts, otherwise we have a bug.